accord-core/src/main/java/accord/utils/RandomSource.java

```diff
@@ -174,6 +178,85 @@ default DoubleStream doubles(double minInclusive, double maxExclusive)
 
     double nextGaussian();
 
+    default int pickInt(int first, int second, int... rest)
+    {
+        int offset = nextInt(0, rest.length + 2);
+        switch (offset)
+        {
+            case 0:  return first;
+            case 1:  return second;
+            default: return rest[offset - 2];
+        }
+    }
+
+    default int pickInt(int[] array)
+    {
+        return pickInt(array, 0, array.length);
+    }
+
+    default int pickInt(int[] array, int offset, int length)
+    {
+        Invariants.checkIndexInBounds(array.length, offset, length);
+        if (length == 1)
+            return array[offset];
+        return array[nextInt(offset, offset + length)];
+    }
+
+    default long pickLong(long first, long second, long... rest)
+    {
+        int offset = nextInt(0, rest.length + 2);
+        switch (offset)
+        {
+            case 0:  return first;
+            case 1:  return second;
+            default: return rest[offset - 2];
+        }
+    }
+
+    default long pickLong(long[] array)
+    {
+        return pickLong(array, 0, array.length);
+    }
+
+    default long pickLong(long[] array, int offset, int length)
+    {
+        Invariants.checkIndexInBounds(array.length, offset, length);
+        if (length == 1)
+            return array[offset];
+        return array[nextInt(offset, offset + length)];
+    }
+
+    default <T extends Comparable<T>> T pick(Set<T> set)
+    {
+        List<T> values = new ArrayList<>(set);
+        values.sort(Comparator.naturalOrder());
```

Comment:
To Benedicts efficiency point, this is probably never going to be efficient without access to the underlying set  so you can randomly probe the array for a value.

still, it's probably faster to just create an iterator and then iterate to a specific index?

Sorting is going to pay that overhead plus more.

Reply:
> Sorting is going to pay that overhead plus more.

without the sort you are bitten by having different ordering on different boxes/jvm, making the selection not really deterministic.  This class of bug has been a pain to fix and find in random testing as it's subtle... had to go fix this pattern in other places so figure it's easier to do here.

there is 1 optimization I don't deal with, if the set is `NavigableSet` I can avoid the sort, but thats not the norm so wouldn't help much